### ISSUE TITLE
rest-core: Make serialization formats extensible by external libraries

### ISSUE THREAD

**Issue by bergmark** [silkapp/rest#20]
#11 adds a new format (Fay JSON) which as is would force rest-core to pull in fay as a dependency. The dependency should be inversed so clients can specify the formats to use and it would prevent dependencies from being added to rest-core for each format.

@boothead You expressed interest in the project at ZuriHac, is this something you still want to work on?

**Comment by hesselink** [comment]
I've been doing some experiments around this a while ago. I haven't worked on it recently, though. Just wanted to say that if anyone is interested in starting on this, my stuff might help a bit.

**Comment by fnoble** [comment]
@hesselink Do you have your experiments up in a branch somewhere? I'd be interested in taking a look and seeing if I could push things forward a bit.

**Comment by hesselink** [comment]
No, it's mostly just exploring the design space for now. I've pushed my work in progress to the [extensible-io branch](https://github.com/silkapp/rest/compare/extensible-io). Expect it to be rebased, though.

The current idea is to represent inputs as a pair of a parser and the accepted content types, and the outputs as a pair of printer and returned content types. The tricky thing I was working on last is that we have derived handlers, and it wasn't clear how to derive those in the new setup. For example, we have multi-PUT and multi-DELETE handlers, where you can batch multiple PUTs into a single API call. This needs a wrapper data type, and that type needs some composed serialization format. Even more complicated is the multi request on top level which can combine any set of API requests together in a single call.

My last experiment was to add a custom existential into the output, and try to cast to that and use it to derive the container serialization for that.

Let me know if you think you can use this and want to work on it, I'd like to be involved in the direction we take this.
